File: hel420.cpp
#include <iostream>
#include <iomanip>
#include <cmath>
#include <vector>
#include <string>
#include "picosha2.h"  // SHA-256 Lib: https://github.com/okdshin/PicoSHA2

// Custom mix function based on a simplified integral formula
double integral_mix_function(double x) {
    if (x == 0.0) x = 1e-10;
    try {
        return pow(x, sin(x)) * pow(63.0, x) + (1.0 / x);
    } catch (...) {
        return pow(x, sin(x)) * 1e10 + (1.0 / x);
    }
}

// Mix the SHA-256 hash using the custom function
std::vector<unsigned char> mix_data(const std::vector<unsigned char>& data, int rounds = 3) {
    std::vector<unsigned char> result = data;

    for (int r = 0; r < rounds; ++r) {
        for (size_t i = 0; i < result.size(); ++i) {
            double x = (static_cast<double>(result[i]) / 255.0) * M_PI;
            double mixed = integral_mix_function(x);
            int byte_val = static_cast<int>(mixed * 1e5) % 256;
            result[i] ^= static_cast<unsigned char>(byte_val);
        }
    }
    return result;
}

// Full HEL-420 hash function
std::string hel_420_hash(const std::string& input) {
    std::vector<unsigned char> hash_bytes(picosha2::k_digest_size);
    picosha2::hash256(input.begin(), input.end(), hash_bytes.begin(), hash_bytes.end());

    std::vector<unsigned char> mixed = mix_data(hash_bytes);

    std::string final_hash;
    picosha2::hash256(mixed.begin(), mixed.end(), final_hash);

    return final_hash;
}


// example main
int main() {
    std::string input = "HEL-420 test message";

    std::string hash_output = hel_420_hash(input);
    std::cout << "HEL-420 hash:\n" << hash_output << std::endl;

    return 0;
}

File: hel420.py
import hashlib
import math
import argparse

def integral_mix_function(x: float) -> float:
    """
    Custom mix function inspired by basic integral logic:
    f(x) = (x^sin(x)) * 63^x + 1/x
    Used to transform byte data in a non-reversible way.
    """
    if x == 0:
        x = 1e-10  # avoid division by zero
    try:
        result = (x ** math.sin(x)) * (63 ** x) + (1 / x)
    except OverflowError:
        # prevent crashing on large powers
        result = (x ** math.sin(x)) * 1e10 + (1 / x)
    return result

def mix_data(data: bytes, rounds: int = 3) -> bytes:
    """
    Applies the mix function to each byte, multiple times.
    """
    result = bytearray(data)
    for _ in range(rounds):
        for i in range(len(result)):
            x = result[i] / 255.0 * math.pi  # scale byte to 0–π
            mixed = integral_mix_function(x)
            result[i] ^= int(mixed * 1e5) % 256
    return bytes(result)

def hel_420_hash(input_data: str) -> str:
    """
    Main HEL-420 hashing function.
    Step 1: SHA-256
    Step 2: Mix with custom transformation
    Step 3: Final SHA-256
    """
    data = input_data.encode('utf-8')
    sha_stage1 = hashlib.sha256(data).digest()
    mixed = mix_data(sha_stage1)
    final_hash = hashlib.sha256(mixed).hexdigest()
    return final_hash

def main():
    parser = argparse.ArgumentParser(description="HEL-420 hash generator (SHA-256 with custom mixing)")
    parser.add_argument("-i", "--input", required=True, help="Input string to hash")
    args = parser.parse_args()

    hash_result = hel_420_hash(args.input)
    print(f"HEL-420 hash:\n{hash_result}")

if __name__ == "__main__":
    main()
